rpio.go

```
@@ -646,11 +668,12 @@ func SetDutyCycle(pin Pin, dutyLen, cycleLen uint32) {
 		shift = 8
 	default:
 		return
+
 	}
 
 	const ctlMask = 255 // ctl setting has 8 bits for each channel
 	const pwen = 1 << 0 // enable pwm
-	const msen = 1 << 7 // use M/S transition instead of pwm algorithm
+	msen = msen << 7
```

The argument should either be bool,
```
if mode == Markspace {
  msen = 1 << 7
}
```
 or you should mask just the single bit
 ```
 msen  = (mode << 7) & (1 << 7)
 ```
  
 Otherwise passing an arbitrary integer > 1 might break something.

---

Good catch, as currently written this is a bug. There is a complication though. If a value other than `Balanced` or `Markspace` is passed it's an error. As currently written the functions don't return an error. To maintain compatibility with existing code, i.e., don't return an error, would require the code to take a default value (e.g., `Markspace`) when provided an invalid value. However, this would be silently masking the error, something I am loath to do.

As a compromise `SetDutyCycle()` can be assumed to always pass a valid value, `Markspace`. `SetDutyCycleMsen()` could be changed to return an error. `SetDutyCycle()` could safely ignore this error. This would maintain backwards compatibility. Any users of `SetDutyCycleMsen()` would be required to handle the error, either by ignoring it (_dangerous_) or explicitly handling it. How do you feel about this approach?

---

Yes, you are right.
Your proposal sounds good.

Or you can do this:
```
const (
  Balanced = false
  Markspace = true
)

func SetDutyCycleSomethingSomething(pin Pin, dutyLen, cycleLen, mode uint32) {
...
const msen = 0
if mode == Markspace {
  msen = 1 << 7
}
...
}
```
That way it would not be possible to enter an invalid value and you don't have to handle the error.

Your solution is also ok though.

---

I keep forgetting this is Go, not C :) I'm going back and forth between the 2 right now.

So, I like your second proposal, changing `Balanced/Markspace` to a `bool` and then have the method signature take a `bool` for the `mode` parameter. I've got to be more aware of which language I'm writing in :)